File: lib/embeddings.js
import { pipeline } from '@xenova/transformers';

let embedder = null;

/**
 * Initialize embedding model (singleton pattern)
 * @returns {Promise} Embedding pipeline instance
 */
export async function initEmbedder() {
  if (embedder) {
    console.log('♻️  Using existing embedder');
    return embedder;
  }

  console.log('🔌 Initializing all-MiniLM-L6-v2 embedder (first run downloads model)...');
  embedder = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');
  console.log('✅ Embedder initialized successfully');
  return embedder;
}

/**
 * Create embeddings for a single text
 * @param {string} text - Text to embed
 * @returns {Promise<Array<number>>} Embedding vector (384 dimensions for all-MiniLM-L6-v2)
 */
export async function createEmbedding(text) {
  const model = await initEmbedder();
  
  console.log(`🤖 [embeddings.js] Creating embedding for text (${text.length} chars)...`);
  
  try {
    const output = await model(text, { pooling: 'mean', normalize: true });
    const embedding = Array.from(output.data);

    console.log(`✅ [embeddings.js] Embedding created (dimensions: ${embedding.length})`);
    return embedding;
  } catch (error) {
    console.error('❌ [embeddings.js] Error creating embedding:', error.message);
    throw error;
  }
}

/**
 * Create embeddings for multiple texts in batch
 * @param {Array<string>} texts - Array of texts to embed
 * @returns {Promise<Array<Array<number>>>} Array of embedding vectors
 */
export async function createEmbeddings(texts) {
  const model = await initEmbedder();
  
  console.log(`🤖 [embeddings.js] Creating embeddings for ${texts.length} texts...`);
  
  const allEmbeddings = [];

  // Process each text individually (local model is fast enough)
  for (let i = 0; i < texts.length; i++) {
    console.log(`📦 [embeddings.js] Processing ${i + 1}/${texts.length}...`);
    
    try {
      const output = await model(texts[i], { pooling: 'mean', normalize: true });
      const embedding = Array.from(output.data);
      allEmbeddings.push(embedding);
      
      console.log(`✅ [embeddings.js] Embedding ${i + 1}/${texts.length} completed`);
    } catch (error) {
      console.error(`❌ [embeddings.js] Error creating embedding for text ${i + 1}:`, error.message);
      throw error;
    }
  }

  console.log(`✅ [embeddings.js] All embeddings created: ${allEmbeddings.length} total`);
  return allEmbeddings;
}

/**
 * Truncate text to fit within token limits
 * Rough estimate: 1 token ≈ 4 characters
 * @param {string} text - Text to truncate
 * @param {number} maxTokens - Maximum tokens (default 8000 for embeddings)
 * @returns {string} Truncated text
 */
export function truncateText(text, maxTokens = 8000) {
  const maxChars = maxTokens * 4; // Rough estimate
  
  if (text.length <= maxChars) {
    return text;
  }

  return text.substring(0, maxChars) + '\n... (truncated)';
}

/**
 * Prepare code text for embedding by cleaning and formatting
 * @param {string} code - Raw code text
 * @param {string} filename - File name/path
 * @returns {string} Prepared text for embedding
 */
export function prepareCodeForEmbedding(code, filename) {
  // Add filename as context
  let prepared = `File: ${filename}\n\n${code}`;
  
  // Truncate if too long
  prepared = truncateText(prepared);
  
  return prepared;
}

/**
 * Prepare documentation text for embedding
 * @param {string} content - Documentation content
 * @param {string} title - Document title or section
 * @returns {string} Prepared text for embedding
 */
export function prepareDocForEmbedding(content, title) {
  let prepared = `Documentation: ${title}\n\n${content}`;
  prepared = truncateText(prepared);
  return prepared;
}

File: python/src/core/embeddings.py
"""
Embedding model operations using sentence-transformers.
Singleton pattern for model initialization, batch embedding support.
"""

import logging
import threading
from typing import List, Optional

from sentence_transformers import SentenceTransformer

from src.config import (
    EMBEDDING_MODEL,
    EMBEDDING_DIMENSIONS,
    MAX_TOKENS,
    CHARS_PER_TOKEN,
)

logger = logging.getLogger(__name__)

# Singleton pattern with thread safety
_embedder: Optional[SentenceTransformer] = None
_lock = threading.Lock()


def get_embedder() -> SentenceTransformer:
    """
    Get or initialize the embedding model (singleton pattern).
    Thread-safe initialization with double-check locking.

    Returns:
        SentenceTransformer: The embedding model instance.
    """
    global _embedder

    if _embedder is not None:
        logger.debug("Using existing embedder")
        return _embedder

    with _lock:
        # Double-check after acquiring lock
        if _embedder is None:
            logger.info(f"Initializing {EMBEDDING_MODEL} embedder (first run downloads model)...")
            _embedder = SentenceTransformer("all-MiniLM-L6-v2")
            logger.info("Embedder initialized successfully")

    return _embedder


def create_embedding(text: str) -> List[float]:
    """
    Create embedding for a single text.

    Args:
        text: Text to embed.

    Returns:
        List of floats representing the embedding vector (384 dimensions).

    Raises:
        Exception: If embedding creation fails.
    """
    model = get_embedder()

    logger.debug(f"Creating embedding for text ({len(text)} chars)...")

    try:
        # sentence-transformers returns numpy array, normalize=True by default
        embedding = model.encode(text, normalize_embeddings=True)
        embedding_list = embedding.tolist()

        logger.debug(f"Embedding created (dimensions: {len(embedding_list)})")

        if len(embedding_list) != EMBEDDING_DIMENSIONS:
            logger.warning(
                f"Unexpected embedding dimensions: {len(embedding_list)} "
                f"(expected {EMBEDDING_DIMENSIONS})"
            )

        return embedding_list
    except Exception as e:
        logger.error(f"Error creating embedding: {e}")
        raise


def create_embeddings(texts: List[str]) -> List[List[float]]:
    """
    Create embeddings for multiple texts in batch.

    Args:
        texts: List of texts to embed.

    Returns:
        List of embedding vectors.

    Raises:
        Exception: If embedding creation fails.
    """
    model = get_embedder()

    logger.info(f"Creating embeddings for {len(texts)} texts...")

    try:
        # sentence-transformers supports batch encoding efficiently
        embeddings = model.encode(texts, normalize_embeddings=True, show_progress_bar=False)
        embeddings_list = [emb.tolist() for emb in embeddings]

        logger.info(f"All embeddings created: {len(embeddings_list)} total")
        return embeddings_list
    except Exception as e:
        logger.error(f"Error creating batch embeddings: {e}")
        raise


def truncate_text(text: str, max_tokens: int = MAX_TOKENS) -> str:
    """
    Truncate text to fit within token limits.
    Rough estimate: 1 token ≈ 4 characters.

    Args:
        text: Text to truncate.
        max_tokens: Maximum tokens (default 8000 for embeddings).

    Returns:
        Truncated text if necessary, otherwise original text.
    """
    max_chars = max_tokens * CHARS_PER_TOKEN

    if len(text) <= max_chars:
        return text

    return text[:max_chars] + "\n... (truncated)"


def prepare_code_for_embedding(code: str, filename: str) -> str:
    """
    Prepare code text for embedding by adding context and formatting.

    Args:
        code: Raw code text.
        filename: File name/path for context.

    Returns:
        Prepared text for embedding.
    """
    prepared = f"File: {filename}\n\n{code}"
    prepared = truncate_text(prepared)
    return prepared


def prepare_doc_for_embedding(content: str, title: str) -> str:
    """
    Prepare documentation text for embedding.

    Args:
        content: Documentation content.
        title: Document title or section name.

    Returns:
        Prepared text for embedding.
    """
    prepared = f"Documentation: {title}\n\n{content}"
    prepared = truncate_text(prepared)
    return prepared
